runtime/yaml.cpp

```diff
@@ -0,0 +1,167 @@
+
+#include "runtime/optional.h"
+#include "runtime/streams.h"
+#include "runtime/critical_section.h"
+#include "runtime/yaml.h"
+
+/*
+ * convert YAML::Node to mixed after parsing a YAML document into YAML::Node
+ */
+static void yaml_node_to_mixed(const YAML::Node &node, mixed &data, const string &source) noexcept {
+  data.clear(); // sets data to NULL
+  if (node.IsScalar()) {
+    const string string_data(node.as<std::string>().c_str());
+    // check whether the primitive is put in quotes in the source YAML
+    const bool string_data_has_quotes = (source[node.Mark().pos] == '"' && source[node.Mark().pos + string_data.size() + 1] == '"');
+    // if so, it is a string
+    if (string_data_has_quotes) {
+      data = string_data;
+    } else if (string_data == string("true")) {
+      data = true; // "true" without quotes is boolean(1)
+    } else if (string_data == string("false")) {
+      data = false; // "false" without quotes is boolean(0)
+    } else if (string_data.is_int()) {
+      data = string_data.to_int();
+    } else {
+      double float_data = 0.0;
+      if (string_data.try_to_float(&float_data)) {
+        data = float_data;
+      } else {
+        data = string_data;
+      }
+    }
+  } else if (node.size() == 0 && node.IsDefined() && !node.IsNull()) {
+    // if node is defined, is not null or scalar and has size 0, then it is an empty array
+    array<mixed> empty_array;
+    data = empty_array;
+  } else if (node.IsSequence()) {
+    for (auto it = node.begin(); it != node.end(); ++it) {
+      mixed data_piece;
+      yaml_node_to_mixed(*it, data_piece, source);
+      data.push_back(data_piece);
+    }
+  } else if (node.IsMap()) {
+    for (const auto &it : node) {
+      mixed data_piece;
+      yaml_node_to_mixed(it.second, data_piece, source);
+      data[string(it.first.as<std::string>().c_str())] = data_piece;
+    }
+  }
+  // else node is Null or Undefined, so data is Null
+}
+
+/*
+ * print tabs in quantity of nesting_level (used to print nested YAML entries)
+ */
+static string yaml_print_tabs(const uint8_t nesting_level) noexcept {
+  return string(2 * nesting_level, ' ');
+}
+
+/*
+ * print the key of a YAML map entry
+ */
+static string yaml_print_key(const mixed& data_key) noexcept {
+  if (data_key.is_string()) {
+    return data_key.as_string();
+  }
+  return string(data_key.as_int()); // key can not be an array; bool and float keys are cast to int
+}
+
+/*
+ * get a YAML representation of mixed in a string variable
+ */
+static void mixed_to_string(const mixed& data, string& string_data, const uint8_t nesting_level = 0) noexcept {
+  string buffer;
+  if (!data.is_array()) {
+    if (data.is_null()) {
+      buffer.push_back('~');
+    } else if (data.is_string()) {
+      const string& string_data_piece = data.as_string();
+      // check if a string has quotes
+      if (string_data_piece.size() < 2 || (string_data_piece[0] != '"' && string_data_piece[string_data_piece.size() - 1] != '"')) {
```

Comment:
This condition seems very strange. Quotes are valid characters. Also, they can occur in the middle of a string. Anyway, they must be escaped